### ISSUE TITLE
How to maintain a uniform sphere of gas?

### ISSUE THREAD

**Issue by trwaters** [PrincetonUniversity/athena-public-version#46]
Hi, I'm trying to get a problem to work in spherical coordinates that currently works fine in cartesian.   In spherical there's either an issue with the inner boundary or I'm missing some trick.  

I can reproduce the issue with the sog (sphere of gas) test attached.  I configure with
`python configure.py --prob=sog  --eos=isothermal --coord=spherical_polar`
[sog_test.zip](https://github.com/PrincetonUniversity/athena-public-version/files/4498708/sog_test.zip)

**Comment by felker** [comment]
Can you attach a plot here?

**Comment by trwaters** [comment]
No matter what boundary condition I try, I'm losing total energy conservation.  
![fig](https://user-images.githubusercontent.com/3180046/79795014-83130a80-8310-11ea-81ea-e23d033c0ba7.png)

**Comment by msbc** [comment]
I think gravity is on by default in spherical polar. I think you need to add `GM = 0` to `<problem>`.

**Comment by trwaters** [comment]
Just tried that.  Same plot as above.

**Comment by ncimerman** [comment]
Hi,

I'm not sure it'll make a difference, but have you tried second order spatial reconstruction to see if it makes a difference? Also have you checked how things change with resolution?

Another thing you could try would be using a different Riemann solver, so e.g. configuring with --solver roe.

From your plots, it also seems that \rho is not 1 initially, as you are trying to set it in the ProblemGenerator. Don't think that this is the issue, but I would change the '1' to '1.0' in line 57 in sog.cpp

Just a few guesses, hope at least one of them helps.

**Comment by trwaters** [comment]
Thanks for these suggestions.  I just tried a few different lower order methods to no avail.  This test problem is definitely sensitive to the domain size and resolution, but this issue remains at higher resolution.

**Comment by felker** [comment]
@msbc gravity is not active, by default

**Comment by trwaters** [comment]
Also, I just verified that cylindrical coordinates handles this problem just fine with outflow BCs.

**Comment by trwaters** [comment]
...no issue either with all of the various versions of my user BC in cylindrical.

**Comment by c-white** [comment]
If you symmetrize the theta bounds about pi/2, e.g. with `x2min = 1.4726215563702154`, `x2max = 1.6689710972195777`, the problem seems to run much better. I can get it to a time of 4 before it blows up.

(Tested with `order = 2`, `integrator = vl2`. Note you were compiling with 2 ghost cells but had 3rd order in the input file. Not sure what the code thinks it's doing in that case.)

**Comment by trwaters** [comment]
Thanks, this was the trick I needed (and I seem to now remember that you pointed out this symmetry breaking to me years ago).  With the llf flux, the solution seems to be maintained indefinitely.

**Comment by pdmullen** [comment]
Hi Tim, 

Good to hear from you; hope you are staying safe/healthy~ 

Can you try the attached athinput ([athinput.sog.txt](https://github.com/PrincetonUniversity/athena-public-version/files/4506509/athinput.sog.txt)) and pgen ([sog.cpp.txt](https://github.com/PrincetonUniversity/athena-public-version/files/4506511/sog.cpp.txt)) with the following configure:

`python configure.py --prob=sog --eos=isothermal --coord=spherical_polar -mpi --flux=roe`

I believe this maintains a uniform sphere of gas to your pre-prescribed `<time>/tlim=100`.  Here is the corresponding movie for the density (https://tinyurl.com/y7khvbcr), which just might be the most exciting media you have seen in all of 2020 thus far.  

Interestingly, I find this test fails without specifying `--flux=roe`.  @felker and @msbc, we have seen some funny behavior with HLLE (isothermal) Riemann solver before, right? (I believe for us it was in the context of STS and general EOS)

Best,
Patrick

**Comment by trwaters** [comment]
Hey Patrick, I'm doing well, hope you are too.  

It is so exciting to see such boring evolution.  Thanks for having a look!  So now I confirm the roe flux also works, but only with the symmetric bounds around pi/2.  Did you use symmetric bounds?

**Comment by trwaters** [comment]
Wait, you're doing full 0 to pi I see.  I got an exception thrown that the origin had to extend to 0 for that.  So your domain here is r = [0, 10], theta = [0,pi] with polar BCs?

**Comment by trwaters** [comment]
Sorry, I just noticed that you included your pgen and input file.  Okay, so I'm in business.  Thank you everyone for your help!

**Comment by kahoooo** [comment]
Glad that it is solved now. But I just took a look into the problem and the underlying problems in the HLLE solver not computing the fluxes correctly.
In lines 121-122 of hydro/rsolvers/hydro/hlle.cpp, `al` and `ar` are redeclared such that the `al` and `ar` in line 72 are not properly initialized.
Removing `Real` from lines 121-122 should fix the problem in case someone wants to use the HLLE solver.

**Comment by trwaters** [comment]
Since this is after all the simplest hydro problem there is, I had to tweet about this particular experience
https://twitter.com/thatTimWaters/status/1252383533479297027?s=20

@kahoooo, good catch!  Although that must not be all, as I'm still not able to get HLLE to work.

**Comment by c-white** [comment]
For completeness, I'm not sure what the history is, but the bug found by @kahoooo isn't in the development branch, so the next public release will fix it.

**Comment by kahoooo** [comment]
@trwaters That's weird since it works for me.
Did you also compile with nghost=3 and symmetrize x2min and x2max about pi/2 as @c-white suggested?
Another note is that user boundary conditions should never touch the active zone although it doesn't break the code in this problem.

**Comment by trwaters** [comment]
Let me try again then, I didn't know you verified it worked.  

Your comment about active zones is interesting because for years I've been setting these in spherical coordinates.  I could not get the code to maintain the Parker and Bondi analytic solutions otherwise.  Perhaps it is related to this...

**Comment by felker** [comment]
@c-white: the change occurred when @msbc reverted the HLLE wavespeed estimates to Roe averages from PVRS on August 26th after the last public release, in:
https://github.com/PrincetonUniversity/athena/commit/1e1a069cb552ca7412e8f200440612baa467b926#diff-557d39248a81dcb3838ed9d3f532444c
https://github.com/PrincetonUniversity/athena/commit/afab25e496b49f2cf7220bdb849e84ce1e86c2cc#diff-557d39248a81dcb3838ed9d3f532444c

I think we should reopen this if only Roe fluxes maintain spherical symmetry and conserve energy for this isothermal sphere of gas in spherical polar coordinates. Unless I am mistaken, LLF and HLLE should also work. @pdmullen, do you have results from those cases handy?

**Comment by pdmullen** [comment]
I just confirmed that my previously attached athinput and pgen combination is stable to `<time>/tlim=100` when using all isothermal Riemann solvers in Athena++ development version.  Therefore, the issue must be isolated to HLLE in public version -- which should be fixed in the next public release.

**Comment by trwaters** [comment]
I also confirmed that my version of sog (with x1min = 1 and x2min/x2max symmetric but not extending to the poles) works for all combinations of vl2, rk2, rk3 and 2nd and 3rd order reconstruction with both Roe and HLLE on the dev version I was granted access to.
